# Ribs puzzle
https://www.quantamagazine.org/the-prime-rib-problem-20170814/

In [5]:
def biggest_chunk(grills, upper_limit = 1000000):
    """
    Finds the biggest chunk (within the upper_limit of ribs) for a given set of grills.
    Accepts a list of grills.

    upper_limit is an optional argument, indicating the maximum snake length (in number of ribs)
        Defaults to 1,000,000
        
    Function is verbose - it prints the results rather than returning them to an object.
    """
        
    consecutive_ribs = 0
    start_rib = 1

    max_consecutive_ribs = 0
    max_start_rib = 0

    for x in range(1, upper_limit + 1):
        if any(x % divisor == 0 for divisor in grills):
            consecutive_ribs += 1
        else:
            if consecutive_ribs > max_consecutive_ribs:
                max_consecutive_ribs = consecutive_ribs
                max_start_rib = start_rib
            else:
                start_rib = start_rib + consecutive_ribs + 1
            start_rib = x + 1
            consecutive_ribs = 0

    print("Biggest chunk was {}.".format(max_consecutive_ribs),
          end = " ")
    print("Chunk was from rib {} to {}.".format(max_start_rib, 
                                               max_start_rib + max_consecutive_ribs - 1))

## Problem 1
Maximum chunk size with five grills

Only prime grills are needed (this is provable).

Heuristic assumptions:
1. Longest chunk size can be found within the first million ribs (dubious - just for computational efficiency).
2. Using the smallest grills (i.e. those with the smallest gaps) would be generate the longest chunks. This will be tested by checking the biggest chunks for all combinations of five primes under 30.

In [6]:
import itertools
import sympy

upper_prime_limit = 30
n_grills = 5

# Get the list of prime numbers under 20
primes_list = list(sympy.sieve.primerange(2, upper_prime_limit))

print("There are {} primes between 1 and {}.".format(len(primes_list), upper_prime_limit), end = " ")
print("There are {} {}-grill combinations.".format(sympy.binomial(len(primes_list), n_grills), n_grills))

# For every combination of x grills, find out the maximum chunk size
for subset in itertools.combinations(primes_list, n_grills):
    print("Grills: {}:".format(subset), end = " ")
    biggest_chunk(subset)

There are 10 primes between 1 and 30. There are 252 5-grill combinations.
Grills: (2, 3, 5, 7, 11): Biggest chunk was 13. Chunk was from rib 114 to 126.
Grills: (2, 3, 5, 7, 13): Biggest chunk was 13. Chunk was from rib 294 to 306.
Grills: (2, 3, 5, 7, 17): Biggest chunk was 13. Chunk was from rib 1134 to 1146.
Grills: (2, 3, 5, 7, 19): Biggest chunk was 13. Chunk was from rib 1344 to 1356.
Grills: (2, 3, 5, 7, 23): Biggest chunk was 13. Chunk was from rib 294 to 306.
Grills: (2, 3, 5, 7, 29): Biggest chunk was 13. Chunk was from rib 1764 to 1776.
Grills: (2, 3, 5, 11, 13): Biggest chunk was 13. Chunk was from rib 864 to 876.
Grills: (2, 3, 5, 11, 17): Biggest chunk was 13. Chunk was from rib 114 to 126.
Grills: (2, 3, 5, 11, 19): Biggest chunk was 13. Chunk was from rib 774 to 786.
Grills: (2, 3, 5, 11, 23): Biggest chunk was 13. Chunk was from rib 3744 to 3756.
Grills: (2, 3, 5, 11, 29): Biggest chunk was 13. Chunk was from rib 1764 to 1776.
Grills: (2, 3, 5, 13, 17): Biggest chunk w

Grills: (2, 7, 13, 17, 19): Biggest chunk was 9. Chunk was from rib 1880 to 1888.
Grills: (2, 7, 13, 17, 23): Biggest chunk was 9. Chunk was from rib 10044 to 10052.
Grills: (2, 7, 13, 17, 29): Biggest chunk was 9. Chunk was from rib 896 to 904.
Grills: (2, 7, 13, 19, 23): Biggest chunk was 9. Chunk was from rib 1078 to 1086.
Grills: (2, 7, 13, 19, 29): Biggest chunk was 9. Chunk was from rib 4610 to 4618.
Grills: (2, 7, 13, 23, 29): Biggest chunk was 9. Chunk was from rib 2870 to 2878.
Grills: (2, 7, 17, 19, 23): Biggest chunk was 9. Chunk was from rib 16 to 24.
Grills: (2, 7, 17, 19, 29): Biggest chunk was 9. Chunk was from rib 1880 to 1888.
Grills: (2, 7, 17, 23, 29): Biggest chunk was 9. Chunk was from rib 896 to 904.
Grills: (2, 7, 19, 23, 29): Biggest chunk was 9. Chunk was from rib 12438 to 12446.
Grills: (2, 11, 13, 17, 19): Biggest chunk was 9. Chunk was from rib 2560 to 2568.
Grills: (2, 11, 13, 17, 23): Biggest chunk was 9. Chunk was from rib 114 to 122.
Grills: (2, 11, 13, 

Grills: (5, 7, 11, 17, 19): Biggest chunk was 6. Chunk was from rib 6950 to 6955.
Grills: (5, 7, 11, 17, 23): Biggest chunk was 6. Chunk was from rib 20535 to 20540.
Grills: (5, 7, 11, 17, 29): Biggest chunk was 6. Chunk was from rib 780 to 785.
Grills: (5, 7, 11, 19, 23): Biggest chunk was 6. Chunk was from rib 3265 to 3270.
Grills: (5, 7, 11, 19, 29): Biggest chunk was 6. Chunk was from rib 8755 to 8760.
Grills: (5, 7, 11, 23, 29): Biggest chunk was 6. Chunk was from rib 780 to 785.
Grills: (5, 7, 13, 17, 19): Biggest chunk was 6. Chunk was from rib 2105 to 2110.
Grills: (5, 7, 13, 17, 23): Biggest chunk was 6. Chunk was from rib 985 to 990.
Grills: (5, 7, 13, 17, 29): Biggest chunk was 6. Chunk was from rib 1390 to 1395.
Grills: (5, 7, 13, 19, 23): Biggest chunk was 6. Chunk was from rib 2505 to 2510.
Grills: (5, 7, 13, 19, 29): Biggest chunk was 6. Chunk was from rib 5015 to 5020.
Grills: (5, 7, 13, 23, 29): Biggest chunk was 6. Chunk was from rib 985 to 990.
Grills: (5, 7, 17, 19,

It would seem that the second assumption is reasonable. The smaller grills generate bigger chunks (for small (< 1m rib) snakes. A {2,3,5,7,11} grill set generates a 13-rib chunk, from a 126-rib snake.

Conjecture: The biggest chunk can be found in a snake whose length is less than the product of the grills. (For example, using grills 2, 3 and 5, we would only need to look at a snake of 2 x 3 x 5 ribs long.) No evidence for this, but the results above do not disprove it.

## Problem 2
Minimum number of grills with 50, 100 and 150-sized chunks.

Heuristic assumptions: Same as problem 1. However, 
1. Longest chunk size can be found within the first million ribs (VERY dubious - just for computational efficiency).
2. Using the smallest grills (i.e. those with the smallest gaps) would be generate the longest chunks. This is somewhat borne out by the results above - but there could be exceptions?

Thus, the most efficient method to achieve a given chunk size would be to keeping adding prime grills. 

In [7]:
import itertools
import sympy

upper_prime_limit = 160
# n_grills = 5

# Get the list of prime numbers under 50
primes_list = list(sympy.sieve.primerange(2, upper_prime_limit))
n_primes = len(primes_list)

print("There are {} primes between 1 and {}.".format(n_primes, upper_prime_limit))

for x in range(1, n_primes+1):
    print("First {} primes: ".format(x), end=" ")
    biggest_chunk(primes_list[:x], 10000000)

There are 37 primes between 1 and 160.
First 1 primes:  Biggest chunk was 1. Chunk was from rib 2 to 2.
First 2 primes:  Biggest chunk was 3. Chunk was from rib 2 to 4.
First 3 primes:  Biggest chunk was 5. Chunk was from rib 2 to 6.
First 4 primes:  Biggest chunk was 9. Chunk was from rib 2 to 10.
First 5 primes:  Biggest chunk was 13. Chunk was from rib 114 to 126.
First 6 primes:  Biggest chunk was 21. Chunk was from rib 9440 to 9460.
First 7 primes:  Biggest chunk was 25. Chunk was from rib 217128 to 217152.
First 8 primes:  Biggest chunk was 33. Chunk was from rib 60044 to 60076.
First 9 primes:  Biggest chunk was 35. Chunk was from rib 8302458 to 8302492.
First 10 primes:  Biggest chunk was 39. Chunk was from rib 9740462 to 9740500.
First 11 primes:  Biggest chunk was 47. Chunk was from rib 7006074 to 7006120.
First 12 primes:  Biggest chunk was 47. Chunk was from rib 7006074 to 7006120.
First 13 primes:  Biggest chunk was 49. Chunk was from rib 7006074 to 7006122.
First 14 prime

For higher numbers of grills, 1m-rib snakes is probably very inadequate, and probably results in trivial findings in many cases. Upping the snake size to 1b ribs indicates that the minimum number of grills needed for a 50-rib chunk would be 12:

In [14]:
upper_limit = 1000000000
n_grills = 12

print("First {} primes: ".format(n_grills), end=" ")

t0 = time.time()
biggest_chunk(primes_list[:n_grills], upper_limit)
t1 = time.time()

elapsed = t1 - t0
print(elapsed)

First 12 primes:  Biggest chunk was 53. Chunk was from rib 132966024 to 132966076.
1252.3240151405334


In [13]:
upper_limit = 1000000000
n_grills = 11

print("First {} primes: ".format(n_grills), end=" ")

t0 = time.time()
biggest_chunk(primes_list[:n_grills], upper_limit)
t1 = time.time()

elapsed = t1 - t0
print(elapsed)

First 11 primes:  Biggest chunk was 49. Chunk was from rib 689448378 to 689448426.
1221.7454769611359
